rust/repositories.bzl

```diff
@@ -42,98 +55,272 @@ filegroup(
-        ("rust-darwin-x86_64", "x86_64-apple-darwin"),
-        ("rust-freebsd-x86_64", "x86_64-unknown-freebsd"),
+def _check_version_valid(version, iso_date, param_prefix = ""):
+    """Verifies that the provided rust version and iso_date make sense."""
+
+    if not version and iso_date:
+        fail("{param_prefix}iso_date must be paired with a {param_prefix}version".format(param_prefix = param_prefix))
+
+    if version in ("beta", "nightly") and not iso_date:
+        fail("{param_prefix}iso_date must be specified if version is 'beta' or 'nightly'".format(param_prefix = param_prefix))
+
+    if version not in ("beta", "nightly") and iso_date:
+        print("{param_prefix}iso_date is ineffective if an exact version is specified".format(param_prefix = param_prefix))
+
+def produce_tool_suburl(tool_name, target_triple, version, iso_date = None):
+    """Produces a fully qualified Rust tool name for URL
+
+    Args:
+      tool_name: The name of the tool per static.rust-lang.org
+      target_triple: The rust-style target triple of the tool
+      version: The version of the tool among "nightly", "beta', or an exact version.
+      iso_date: The date of the tool (or None, if the version is a specific version).
+    """
+
+    if iso_date:
+        return "{}/{}-{}-{}".format(iso_date, tool_name, version, target_triple)
+    else:
+        return "{}-{}-{}".format(tool_name, version, target_triple)
+
+def produce_tool_path(tool_name, target_triple, version):
+    """Produces a qualified Rust tool name
+
+    Args:
+      tool_name: The name of the tool per static.rust-lang.org
+      target_triple: The rust-style target triple of the tool
+      version: The version of the tool among "nightly", "beta', or an exact version.
+    """
+
+    return "{}-{}-{}".format(tool_name, version, target_triple)
+
+def load_arbitrary_tool(ctx, tool_name, param_prefix, tool_subdirectory, version, iso_date, target_triple):
+    """Loads a Rust tool, downloads, and extracts into the common workspace.
+
+    Args:
+      ctx: A repository_ctx.
+      tool_name: The name of the given tool per the archive naming.
+      param_prefix: The name of the versioning param if the repository rule supports multiple tools.
+      tool_subdirectory: The subdirectory of the tool files (wo level below the root directory of
+                         the archive. The root directory of the archive is expected to match
+                         $TOOL_NAME-$VERSION-$TARGET_TRIPLE.
+      version: The version of the tool among "nightly", "beta', or an exact version.
+      iso_date: The date of the tool (or None, if the version is a specific version).
+      target_triple: The rust-style target triple of the tool
+    """
+
+    _check_version_valid(version, iso_date, param_prefix)
+
+    tool_suburl = produce_tool_suburl(tool_name, target_triple, version, iso_date)
+    url = "https://static.rust-lang.org/dist/{}.tar.gz".format(tool_suburl)
```

Comment:
some breadcrumbs to https://static.rust-lang.org/dist/index.html may be useful

Reply:
Done in docstring + above the url decl.